File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
// @Author: KING-SEN

class Solution {
public:
    bool IsPossible(vector<int>& piles, int k, int h) {
        int len = piles.size();
        long long hours = 0;
        for (int i = 0; i < len; i++) {
            hours += (piles[i] / k) + (piles[i] % k != 0 ? 1 : 0);
        }
        return hours <= (1ll * h); 
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int start = 1, end = 1e9, answer = 0;
        while (start <= end) {
            int mid = (end - start) / 2 + start;
            if (IsPossible(piles, mid, h)) {
                answer = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return answer;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def feasible(speed) -> bool:
            return sum(math.ceil(pile / speed) for pile in piles) <= h
        
        low, high, ans = 1, max(piles), -1
        while low <= high:
            mid = (low + high) // 2
            if feasible(mid):
                ans = mid
                high = mid - 1
            else:
                low = mid + 1
        return ans